pkg/filesystem/subvolume.go

```diff
@@ -0,0 +1,155 @@
+}
+
+func List(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStale bool) map[string]subVolumeInfo {
+
+	if !includeStale {
+		logging.Info("Listing subvolumes")
+	} else {
+		logging.Info("Listing stale subvolumes")
+	}
+	subvolumeNames := getK8sRefSubvolume(ctx, clientsets)
+	list := listCephFSSubvolumes(ctx, clientsets, operatorNamespace, clusterNamespace, includeStale, subvolumeNames)
+	if len(list) == 0 && includeStale {
+		logging.Info("No stale subvolume found")
+	}
+	return list
+}
+
+// getk8sRefSubvolume returns the k8s ref for the subvolumes
+func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map[string]subVolumeInfo {
+	pvList, err := clientsets.Kube.CoreV1().PersistentVolumes().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching PVs: %v\n", err))
+		os.Exit(1)
+	}
+	subvolumeNames := make(map[string]subVolumeInfo)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI.VolumeAttributes["subvolumeName"] != "" {
+			subvolumeNames[pv.Spec.CSI.VolumeAttributes["subvolumeName"]] = subVolumeInfo{}
+		}
+	}
+	return subvolumeNames
+}
+
+// listCephFSSubvolumes list all the subvolumes
+func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStale bool, subvolumeNames map[string]subVolumeInfo) map[string]subVolumeInfo {
+
+	// getFilesystem gets the filesystem
+	fsstruct := getFileSystem(ctx, clientsets, operatorNamespace, clusterNamespace)
+	staleSubvolumeNames := make(map[string]subVolumeInfo)
+	var svList string
+
+	// this iterates over the filesystems and subvolumegroup to get the list of subvolumes that exist
+	for _, fs := range fsstruct {
+		// gets the subvolumegroup in the filesystem
+		subvolg := getSubvolumeGroup(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name)
+		for _, svg := range subvolg {
+			svList = exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "ls", fs.Name, svg.Name}, operatorNamespace, clusterNamespace, true, false)
+			subvol := unMarshaljson(svList)
+			if len(subvol) == 0 {
+				logging.Info("No subvolume found for filesystem %s and subvolumegroup %s", fs, svg)
+				continue
+			}
+
+			// append the subvolume which doesn't have any snapshot attached to it.
+			for _, sv := range subvol {
+				if includeStale == true {
+					// lookup for subvolume in list of the k8s Ref, if not it can be counted as stale subvolume
+					_, ok := subvolumeNames[sv.Name]
+					subvolumeNames[sv.Name] = subVolumeInfo{fs.Name, svg.Name}
+					if !ok {
+						if !checkSnapshot(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name, sv.Name, svg.Name) {
+							staleSubvolumeNames[sv.Name] = subVolumeInfo{fs.Name, svg.Name}
+							logging.Info("%s %s %s", fs.Name, sv.Name, svg.Name)
+						} else {
+							logging.Info("subvolume %s has snapshots", sv.Name)
+						}
+					}
+				} else {
```

Comment:
If you have the `continue` above, then you can always log this, and you can remove the duplicated log output on line 101 above. It just keeps the log output in a single place.
```suggestion
				}
```